prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -1935,11 +1935,17 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                         }
                         tymap_stack.push(tymap);
                     }
-                    let cleanup = |this: &ProcedureEncoder| {
-                        // FIXME: this is a hack to support generics. See issue #187.
-                        let mut tymap_stack = this.encoder.typaram_repl.borrow_mut();
-                        tymap_stack.pop();
-                    };
+
+                    // FIXME: this is a hack to support generics. See issue #187.
+                    struct CleanupTyMapStack<'a, 'tcx> {
```

Comment:
It would be a bit nicer if this was defined only once.

Reply:
true. i think it would make sense then to have some method like `enter` or `push` that enters the new "tymap stack frame" and returns the cleanup token, but i'll leave that for some other time i guess and mark this PR as Draft.

Reply:
found some time today to implement this with a `push_temp_tymap` method on `Encoder`, so we only have the struct and drop impl once. what do you think?

Reply:
I guess it makes sense that the API is on `Encoder`. Maybe the name is not as self-explanatory as it could be but then neither is `tymap` itself. This is an improvement overall, if CI passes I'm happy to merge this.